### PR TITLE
Deny missing Debug and Copy from data structures

### PR DESCRIPTION
**Pull request by mfpiccolo**
You will see a few TODO's because the fields down impl copy or debug.

### CODE REVIEW

File: diesel/src/expression/array_comparison.rs [hunk 1]
```diff
@@ -100,6 +101,8 @@ impl<ST, S, F, W, O, L, Of> AsInExpression<ST>
     }
 }
 
+// TODO: Copy
```
Comment (sgrif) [inline diesel/src/expression/array_comparison.rs]: This is not a TODO, `Many<T>` should never `impl Copy`.

File: diesel/src/pg/backend.rs [hunk 2]
```diff
@@ -2,6 +2,7 @@ use backend::*;
 use query_builder::bind_collector::RawBytesBindCollector;
 use super::query_builder::PgQueryBuilder;
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/backend.rs]: You don't think it makes sense to impl these traits here?

File: diesel/src/pg/connection/mod.rs [hunk 3]
```diff
@@ -26,6 +26,7 @@ use types::{ToSql, HasSqlType};
 /// The connection string expected by `PgConnection::establish`
 /// should be a PostgreSQL connection string, as documented at
 /// http://www.postgresql.org/docs/9.4/static/libpq-connect.html#LIBPQ-CONNSTRING
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/connection/mod.rs]: I don't think the `missing_copy_implementation` needs to be explicitly allowed here, does it?

File: diesel/src/pg/expression/array_comparison.rs [hunk 4]
```diff
@@ -46,6 +46,7 @@ pub fn any<ST, T>(vals: T) -> Any<T::Expression, ST> where
 }
 
 #[doc(hidden)]
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/expression/array_comparison.rs]: This type should probably implement these?

File: diesel/src/pg/expression/date_and_time.rs [hunk 5]
```diff
@@ -5,6 +5,7 @@ use query_builder::*;
 use result::QueryResult;
 use types::{Timestamp, VarChar};
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/expression/date_and_time.rs]: This type should probably implement these?

File: diesel/src/pg/query_builder.rs [hunk 6]
```diff
@@ -4,6 +4,7 @@ use super::backend::Pg;
 use super::connection::raw::RawConnection;
 use query_builder::{QueryBuilder, BuildQueryResult};
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/query_builder.rs]: I don't think we need to explicitly allow `missing_copy_implementations` here.

File: diesel/src/pg/types/mod.rs [hunk 7]
```diff
@@ -8,8 +8,12 @@ mod uuid;
 
 #[doc(hidden)]
 pub mod sql_types {
-    #[derive(Clone, Copy, Default)] pub struct Oid;
-    #[derive(Clone, Copy, Default)] pub struct Array<T>(T);
+    #[allow(missing_debug_implementations)]
+    #[derive(Clone, Copy, Default)]
+    pub struct Oid;
+    #[allow(missing_debug_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/types/mod.rs]: This should probably implement `Debug`?

File: diesel/src/pg/types/mod.rs [hunk 8]
```diff
@@ -8,8 +8,12 @@ mod uuid;
 
 #[doc(hidden)]
 pub mod sql_types {
-    #[derive(Clone, Copy, Default)] pub struct Oid;
-    #[derive(Clone, Copy, Default)] pub struct Array<T>(T);
+    #[allow(missing_debug_implementations)]
```
Comment (sgrif) [inline diesel/src/pg/types/mod.rs]: This should probably implement `Debug`?

File: diesel/src/query_builder/bind_collector.rs [hunk 9]
```diff
@@ -5,6 +5,8 @@ pub trait BindCollector<DB: Backend> {
     fn push_bound_value<T>(&mut self, bind: Option<Vec<u8>>) where DB: HasSqlType<T>;
 }
 
+// TODO: copy
```
Comment (sgrif) [inline diesel/src/query_builder/bind_collector.rs]: This type would never implement `Copy`

File: diesel/src/query_builder/delete_statement.rs [hunk 10]
```diff
@@ -2,6 +2,7 @@ use backend::Backend;
 use query_builder::*;
 use result::QueryResult;
 
+#[derive(Debug)]
```
Comment (sgrif) [inline diesel/src/query_builder/delete_statement.rs]: Isn't this one missing `Copy`?

File: diesel/src/query_builder/select_statement/boxed.rs [hunk 11]
```diff
@@ -11,6 +11,8 @@ use query_source::QuerySource;
 use result::QueryResult;
 use types::{HasSqlType, Bool, BigInt};
 
+// TODO: Copy Debug
```
Comment (sgrif) [inline diesel/src/query_builder/select_statement/boxed.rs]: This type would never implement either one of these.

File: diesel/src/query_builder/select_statement/boxed.rs [hunk 12]
```diff
@@ -11,6 +11,8 @@ use query_source::QuerySource;
...
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/query_builder/select_statement/boxed.rs]: I don't think this needs to be explicitly allowed?

File: diesel/src/sqlite/backend.rs [hunk 13]
```diff
@@ -3,8 +3,10 @@ use query_builder::bind_collector::RawBytesBindCollector;
 use super::connection::SqliteValue;
 use super::query_builder::SqliteQueryBuilder;
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/sqlite/backend.rs]: This type should probably implement these?

File: diesel/src/sqlite/connection/mod.rs [hunk 14]
```diff
@@ -28,6 +28,7 @@ use sqlite::Sqlite;
 use super::query_builder::SqliteQueryBuilder;
 use types::HasSqlType;
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/sqlite/connection/mod.rs]: Copy shouldn't need to be explicitly allowed.

File: diesel/src/sqlite/query_builder/mod.rs [hunk 15]
```diff
@@ -5,6 +5,7 @@ pub mod functions;
 #[doc(hidden)]
 pub mod nodes;
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/sqlite/query_builder/mod.rs]: Copy shouldn't need to be explicitly allowed.

File: diesel/src/sqlite/query_builder/nodes.rs [hunk 16]
```diff
@@ -3,6 +3,7 @@ use query_builder::{QueryFragment, QueryBuilder, BuildQueryResult};
 use result::QueryResult;
 use sqlite::Sqlite;
 
+#[allow(missing_debug_implementations, missing_copy_implementations)]
```
Comment (sgrif) [inline diesel/src/sqlite/query_builder/nodes.rs]: This should probably implement these?

**Comment by mfpiccolo** [comment]
@sgrif I believe comments have been addressed.  Still need to look into compile test errors.

**Comment by sgrif** [comment]
Doesn't look like there are any compile test errors?

**Comment by mfpiccolo** [comment]
👍 Thats weird.  I had them locally.  Must be my nightly version.

**Comment by sgrif** [comment]
Yeah, make sure you're using the same version in `.travis.yml`. I'll bump it soon as there are breaking changes coming in libsyntax tomorrow. LGTM

**Comment by mfpiccolo** [comment]
Thanks for the help! 😄

**Pull request merged**